File: app/src/main/java/de/devmil/muzei/bingimageofthedayartsource/LicenseInfoActivity.java
/*
 * Copyright 2014 Devmil Solutions
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package de.devmil.muzei.bingimageofthedayartsource;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Typeface;
import android.net.Uri;
import android.os.Bundle;
import android.text.Html;
import android.text.util.Linkify;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.ScrollView;
import android.widget.TextView;

import de.devmil.common.licensing.LicenseManager;
import de.devmil.common.licensing.PackageInfo;

import static de.devmil.common.utils.LogUtil.LOGW;

public class LicenseInfoActivity extends Activity {

    private static final String TAG = LicenseInfoActivity.class.getSimpleName();

    private LicenseManager _LicenseManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_license_info);

        _LicenseManager = new LicenseManager(this, R.raw.licenseinfo);

        ListView listView = (ListView)findViewById(R.id.activity_license_info_listView);

        LicenseEntryAdapter adapter = new LicenseEntryAdapter(this, _LicenseManager.getLicenseInfo().getPackages().toArray(new PackageInfo[0]));

        listView.setAdapter(adapter);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if(item.getItemId() == android.R.id.home)
        {
            setResult(Activity.RESULT_OK);
            finish();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    class LicenseEntryAdapter extends ArrayAdapter<PackageInfo>
    {
        public LicenseEntryAdapter(Context context, PackageInfo[] items) {
            super(context, 0, items);
        }

        class ViewHolder
        {
            ImageView image;
            TextView name;
            TextView copyright;
            TextView url;
            Button licenseButton;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            if(convertView == null)
            {
                convertView = LayoutInflater.from(getContext()).inflate(R.layout.activity_license_info_entry, null);
                ViewHolder vh = new ViewHolder();
                vh.image = (ImageView)convertView.findViewById(R.id.activity_license_info_entry_image);
                vh.name = (TextView)convertView.findViewById(R.id.activity_license_info_entry_name);
                vh.copyright = (TextView)convertView.findViewById(R.id.activity_license_info_entry_copyright);
                vh.url = (TextView)convertView.findViewById(R.id.activity_license_info_entry_url);
                vh.licenseButton = (Button)convertView.findViewById(R.id.activity_license_info_entry_licensebutton);
                convertView.setTag(vh);
            }

            ViewHolder holder = (ViewHolder)convertView.getTag();

            PackageInfo item = getItem(position);

            int drawableId = getContext().getResources().getIdentifier(item.getIconName(), "drawable", "de.devmil.muzei.bingimageofthedayartsource");

            holder.image.setImageResource(drawableId);
            holder.name.setText(item.getName());
            holder.copyright.setText(item.getCopyright());
            holder.url.setText(Html.fromHtml("<a href=\"" + item.getUrl() + "\">" + item.getUrl() + "</a>"));
            holder.url.setAutoLinkMask(Linkify.WEB_URLS);
            holder.url.setTag(item);
            holder.url.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    try {
                        PackageInfo pi = (PackageInfo) v.getTag();
                        Intent intent = new Intent(Intent.ACTION_VIEW);
                        intent.setData(Uri.parse(pi.getUrl()));

                        startActivity(intent);
                    }
                    catch(Exception e)
                    {
                        LOGW(TAG, "Error parsing package url", e);
                    }
                }
            });
            holder.licenseButton.setText(item.getLicense().getName());
            holder.licenseButton.setTag(item);
            holder.licenseButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    PackageInfo pi = (PackageInfo) v.getTag();
                    ScrollView scrollView = new ScrollView(getContext());
                    TextView tvMessage = new TextView(getContext());
                    tvMessage.setTextSize(TypedValue.COMPLEX_UNIT_DIP, 6);
                    tvMessage.setTypeface(Typeface.MONOSPACE);
                    tvMessage.setText(pi.getLicense().getContent());
                    scrollView.addView(tvMessage);
                    AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
                    AlertDialog dlg = builder.setTitle(pi.getLicense().getName())
                            .setView(scrollView)
                            .setNegativeButton("OK", new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialog, int which) {
                                    dialog.dismiss();
                                }
                            })
                            .create();
                    dlg.show();

                }
            });

            return convertView;
        }
    }
}

File: app/src/main/java/de/devmil/muzei/bingimageoftheday/LicenseInfoActivity.kt
/*
 * Copyright 2014 Devmil Solutions
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package de.devmil.muzei.bingimageoftheday

import android.annotation.SuppressLint
import android.app.Activity
import android.app.AlertDialog
import android.content.Context
import android.content.Intent
import android.graphics.Typeface
import android.net.Uri
import android.os.Bundle
import android.text.Html
import android.text.util.Linkify
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.*
import de.devmil.common.licensing.LicenseManager
import de.devmil.common.licensing.PackageInfo
import de.devmil.common.utils.LogUtil

class LicenseInfoActivity : Activity() {

    private var licenseManager: LicenseManager? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_license_info)

        licenseManager = LicenseManager(this, R.raw.licenseinfo)

        val listView = findViewById(R.id.activity_license_info_listView) as ListView

        val adapter = LicenseEntryAdapter(this, licenseManager!!.licenseInfo!!.packages.toTypedArray())

        listView.adapter = adapter
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            setResult(Activity.RESULT_OK)
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    internal inner class LicenseEntryAdapter(context: Context, items: Array<PackageInfo>) : ArrayAdapter<PackageInfo>(context, 0, items) {

        internal inner class ViewHolder {
            var image: ImageView? = null
            var name: TextView? = null
            var copyright: TextView? = null
            var url: TextView? = null
            var licenseButton: Button? = null
        }

        @SuppressLint("InflateParams")
        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            var effectiveconvertView = convertView
            if (effectiveconvertView == null) {
                effectiveconvertView = LayoutInflater.from(context).inflate(R.layout.activity_license_info_entry, null)
                val vh = ViewHolder()
                vh.image = effectiveconvertView!!.findViewById(R.id.activity_license_info_entry_image) as ImageView
                vh.name = effectiveconvertView.findViewById(R.id.activity_license_info_entry_name) as TextView
                vh.copyright = effectiveconvertView.findViewById(R.id.activity_license_info_entry_copyright) as TextView
                vh.url = effectiveconvertView.findViewById(R.id.activity_license_info_entry_url) as TextView
                vh.licenseButton = effectiveconvertView.findViewById(R.id.activity_license_info_entry_licensebutton) as Button
                effectiveconvertView.tag = vh
            }

            val holder = effectiveconvertView.tag as ViewHolder

            val item = getItem(position)

            //context.getResources().getIdentifier() doesn't work any more!?
            val drawableId = R.drawable::class.java.getField(item!!.iconName).getInt(null)

            holder.image!!.setImageResource(drawableId)
            holder.name!!.text = item.name
            holder.copyright!!.text = item.copyright
            @Suppress("DEPRECATION")
            holder.url!!.text = Html.fromHtml("<a href=\"" + item.url + "\">" + item.url + "</a>")
            holder.url!!.autoLinkMask = Linkify.WEB_URLS
            holder.url!!.tag = item
            holder.url!!.setOnClickListener { v ->
                try {
                    val pi = v.tag as PackageInfo
                    val intent = Intent(Intent.ACTION_VIEW)
                    intent.data = Uri.parse(pi.url)

                    startActivity(intent)
                } catch (e: Exception) {
                    LogUtil.LOGW(TAG, "Error parsing package url", e)
                }
            }
            holder.licenseButton!!.text = item.license.name
            holder.licenseButton!!.tag = item
            holder.licenseButton!!.setOnClickListener { v ->
                val pi = v.tag as PackageInfo
                val scrollView = ScrollView(context)
                val tvMessage = TextView(context)
                tvMessage.setTextSize(TypedValue.COMPLEX_UNIT_DIP, 6f)
                tvMessage.typeface = Typeface.MONOSPACE
                tvMessage.text = pi.license.content
                scrollView.addView(tvMessage)
                val builder = AlertDialog.Builder(context)
                val dlg = builder.setTitle(pi.license.name)
                        .setView(scrollView)
                        .setNegativeButton("OK") { dialog, _ -> dialog.dismiss() }
                        .create()
                dlg.show()
            }

            return effectiveconvertView
        }
    }

    companion object {

        private val TAG = LicenseInfoActivity::class.java.simpleName
    }
}
